From b2e04a04753979e44f6c868afd0335077fb605ba Mon Sep 17 00:00:00 2001
From: Matyas Selmeci <matyas@cs.wisc.edu>
Date: Tue, 3 May 2016 16:13:58 -0500
Subject: [PATCH] SOFTWARE-2288 Auto-create server keys

If explicitly asked to via the AUTOCREATE_SERVER_KEYS variable in
/etc/sysconfig/gsisshd, generate RSA, DSA, and ECDSA SSH host keys in
the init script.

AUTOCREATE_SERVER_KEYS must be "YES" or "RSAONLY" to generate keys.
---
 gsisshd.init | 37 +++++++++++++++++++++++++++++--------
 1 file changed, 29 insertions(+), 8 deletions(-)

diff --git a/gsisshd.init b/gsisshd.init
index bdbb94c..39b8b2e 100755
--- a/gsisshd.init
+++ b/gsisshd.init
@@ -41,10 +41,11 @@ lockfile=/var/lock/subsys/$prog
 KEYGEN=/usr/bin/gsissh-keygen
 SSHD=/usr/sbin/gsisshd
 RSA1_KEY=/etc/gsissh/ssh_host_key
 RSA_KEY=/etc/gsissh/ssh_host_rsa_key
 DSA_KEY=/etc/gsissh/ssh_host_dsa_key
+ECDSA_KEY=/etc/gsissh/ssh_host_ecdsa_key
 PID_FILE=/var/run/gsisshd.pid
 
 runlevel=$(set -- $(runlevel); eval "echo \$$#" )
 
 do_rsa1_keygen() {
@@ -105,10 +106,30 @@ do_dsa_keygen() {
 			exit 1
 		fi
 	fi
 }
 
+do_ecdsa_keygen() {
+	if [ ! -s $ECDSA_KEY ]; then
+		echo -n $"Generating SSH2 ECDSA host key: "
+		rm -f $ECDSA_KEY
+		if test ! -f $ECDSA_KEY && $KEYGEN -q -t ecdsa -f $ECDSA_KEY -C '' -N '' >&/dev/null; then
+			chmod 600 $ECDSA_KEY
+			chmod 644 $ECDSA_KEY.pub
+			if [ -x /sbin/restorecon ]; then
+			    /sbin/restorecon $ECDSA_KEY.pub
+			fi
+			success $"ECDSA key generation"
+			echo
+		else
+			failure $"ECDSA key generation"
+			echo
+			exit 1
+		fi
+	fi
+}
+
 do_restart_sanity_check()
 {
 	$SSHD -t
 	RETVAL=$?
 	if [ $RETVAL -ne  0 ]; then
@@ -120,18 +141,18 @@ do_restart_sanity_check()
 start()
 {
 	[ -x $SSHD ] || exit 5
 	[ -f /etc/gsissh/sshd_config ] || exit 6
 	# Create keys if necessary
-	# - Not necessary if running with keyex only
-	# if [ "x${AUTOCREATE_SERVER_KEYS}" != xNO ]; then
-	#	do_rsa_keygen
-	#	if [ "x${AUTOCREATE_SERVER_KEYS}" != xRSAONLY ]; then
-	#		do_rsa1_keygen
-	#		do_dsa_keygen
-	#	fi
-	# fi
+	if [ "x${AUTOCREATE_SERVER_KEYS}" = xRSAONLY ]; then
+		do_rsa_keygen
+	elif [ "x${AUTOCREATE_SERVER_KEYS}" = xYES ]; then
+		do_rsa_keygen
+		do_rsa1_keygen
+		do_dsa_keygen
+		do_ecdsa_keygen
+	fi
 
 	echo -n $"Starting $prog: "
 	$SSHD $OPTIONS && success || failure
 	RETVAL=$?
 	[ $RETVAL -eq 0 ] && touch $lockfile
-- 
2.6.3

